### PR TITLE
Creating tensor class

### PR DESCRIPTION
**Pull request by sayandip18**

### CODE REVIEW

File: sparse/ndsparse/tensor.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import pytaco as pt
+from pytaco import compressed
+import numpy as np
+from numpy.lib.mixins import NDArrayOperatorsMixin
+
+
+class tensor(np.ndarray, pt.tensor, NDArrayOperatorsMixin):
```
Comment (hameerabbasi) [inline sparse/ndsparse/tensor.py:7]: We should wrap `pt.tensor`, not inherit from it. And `np.ndarray` shouldn't be involved at all.

**Review by hameerabbasi — COMMENTED** [review]
Overall, this seems to be in the right direction.

**Comment by sayandip18** [comment]
Note: The docstring in the latest commit has been copypasted from the Pytaco docstring as I didn't want to switch tabs every time :grimacing: 
 The docstring will be changed in the next commit.

**Comment by sayandip18** [comment]
Todo: modify the dunder methods, add test suite, get taco in the CI environment so that tests pass.

File: sparse/ndsparse/_tensor.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    np.int16,
+    np.int32,
+    np.int64,
+    np.uint8,
+    np.uint16,
+    np.uint32,
+    np.uint64,
+    np.float16,
+    np.float32,
+]
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:22]: This is not needed, just use NumPy types directly.

File: sparse/ndsparse/_tensor.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    name
+    shape
+    format
+    dtype
+    T
+
+    """
+
+    def __init__(self, coords, data=None, dtype=float, name=None):
+        if isinstance(name, str) or name is None:
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:52]: We don't need a name for the tensor.

File: sparse/ndsparse/_tensor.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            self.coords = coords
+
+        elif isinstance(coords, str):
+            raise ValueError("String is not a valid input")
+
+        elif hasattr(coords, "__iter__"):
+            self.coords = coords
+
+        else:
+            raise ValueError("Not an iterable object")
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:67]: Should be `shape`

File: sparse/ndsparse/_tensor.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        return pt.tensor_abs(self, self.format)
+
+    def __neg__(self):
+        return pt.tensor_neg(self, self.format)
+
+    def __eq__(self, other):
+        return pt.tensor_eq(self, other, mode)
+
+    def __neg__(self):
+        return pt.tensor_neg(self, self.format)
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:204]: This is taken care of by `NDArrayOperatorsMixin` and you can override `__array_ufunc__` instead.

**Review by hameerabbasi — COMMENTED** [review]
Couple of minor comments

**Comment by hameerabbasi** [comment]
We still need to figure out how to build `pytaco` as part of building `sparse`, instead of doing it manually.

**Comment by sayandip18** [comment]
Yes, I'll give it a try today.

File: sparse/ndsparse/_tensor.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            elif isinstance(arg, SparseArray):
+                arg_converted.append(np.asarray(arg.todense()))
+            elif isinstance(arg, np.ndarray):
+                arg_converted.append(arg)
+            elif isinstance(arg, sparse.tensor):
+                arg_converted.append(arg.todense())
+            elif isinstance(arg, scipy.sparse.spmatrix):
+                arg_converted.append(arg.toarray())
+            else:
+                raise TypeError("Incompatible type")
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:171]: We should round-tripping from dense tensors.

File: sparse/ndsparse/_tensor.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        """
+        return self._tensor.order()
+
+    def transpose(self, new_ordering, new_format=None, name=None):
+        """
+        Returns the transpose of the tensor
+        """
+        return self._tensor.transpose(new_ordering, new_format, name)
+
+    def T():
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:100]:
```suggestion
    @propperty
    def T(self):
```

File: sparse/ndsparse/_tensor.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        elif isinstance(shape, str):
+            raise ValueError("String is not a valid input")
+
+        elif hasattr(shape, "__iter__"):
+            self.shape = shape
+
+        else:
+            raise ValueError("Not an iterable object")
+
+        self._tensor = pt.tensor(shape)
```
Comment (hameerabbasi) [inline sparse/ndsparse/_tensor.py:45]: Use super constructor.

**Pull request closed without merging**